src/valid/handles.rs

```diff
@@ -0,0 +1,337 @@
+//! Implementation of [`super::Validator::validate_module_handles`].
+
+use crate::{arena::BadHandle, Arena, Handle, UniqueArena};
+use std::{convert::TryInto, hash::Hash, num::NonZeroU32};
+
+impl super::Validator {
+    /// Validates that all handles within `module` are:
+    ///
+    /// * Valid, in the sense that they contain indices within each arena structure inside the
+    /// [`crate::Module`] type.
+    /// * No arena contents contain any items that have forward dependencies; that is, the value
+    ///     associated with a handle only may contain references to handles in the same arena that
+    ///     were constructed before it.
+    ///
+    /// By validating the above conditions, we free up subsequent logic to assume that handle
+    /// accesses are infallible.
+    ///
+    /// # Errors
+    ///
+    /// Errors returned by this method are intentionally sparse, for simplicity of implementation.
+    /// It is expected that only buggy frontends or fuzzers should ever emit IR that fails this
+    /// validation pass.
+    pub(super) fn validate_module_handles(
+        module: &crate::Module,
+    ) -> Result<(), InvalidHandleError> {
+        let &crate::Module {
+            ref constants,
+            ref entry_points,
+            ref functions,
+            ref global_variables,
+            ref types,
+        } = module;
+
+        // NOTE: Types being first is important. All other forms of validation depend on this.
+        for (this_handle, type_) in types.iter() {
+            let &crate::Type { name: _, ref inner } = type_;
+
+            match inner {
+                &crate::TypeInner::Scalar { .. }
+                | &crate::TypeInner::Vector { .. }
+                | &crate::TypeInner::Matrix { .. }
+                | &crate::TypeInner::ValuePointer { .. }
+                | &crate::TypeInner::Atomic { .. }
+                | &crate::TypeInner::Image { .. }
+                | &crate::TypeInner::Sampler { .. } => (),
+                &crate::TypeInner::Pointer { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Array { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Struct { ref members, .. } => {
+                    this_handle.check_dep_iter(members.iter().map(|m| m.ty))?;
+                }
+                &crate::TypeInner::BindingArray { base, .. } => {
```

Comment:
Everything that applies to `Array` sizes applies here as well, I think.

Reply:
Since you're using [a reference to another convo](https://github.com/gfx-rs/naga/pull/2090#discussion_r1044946235)...here's [a reference to my answer there](https://github.com/gfx-rs/naga/pull/2090#discussion_r1047294521). 😉